### PR TITLE
refactor(messages): extract basic factories (5 of 8)

### PR DESCRIPTION
**Pull request by jatmn**
## Summary

Part 5 of 8 in the messages.ts de-monolith migration.

This standalone PR extracts basic message factories and adjacent synthetic-message helpers from src/utils/messages.ts into src/utils/messages/factories.ts. The public messages.ts surface continues to re-export the moved APIs so existing imports remain valid while this migration lands one PR at a time.

Associated test coverage included:
- Added src/utils/messages/factories.test.ts for the moved factory helpers.
- Added coverage for the reviewer-raised legacy non-array assistant-content edge case in hasToolCallsInLastAssistantTurn.

## Related active PR impact

No direct active PR overlap found for this factory slice. #1614 still directly touches src/utils/messages.ts, but this PR avoids the normalization/tool-id areas that #1614 edits.

## Validation

- bun install --frozen-lockfile
- bun test src/utils/messages/factories.test.ts
- bun test --feature=UNATTENDED_RETRY --max-concurrency=1 src/services/compact/compact.test.ts src/services/tools/toolExecution.test.ts src/services/tools/StreamingToolExecutor.test.ts src/services/goal/controller.test.ts
- bun test src/utils/messages/factories.test.ts src/services/compact/compact.test.ts src/services/tools/toolExecution.test.ts src/services/tools/StreamingToolExecutor.test.ts src/services/goal/controller.test.ts
- bun test src/query.abortClassification.test.ts src/query/stopHooks.goal.test.ts
- bun test src/services/tools/toolExecution.test.ts
- bun test src/commands/branch/branch.test.ts src/services/awaySummary.test.ts
- bun test src/services/goal/controller.test.ts
- bun run typecheck
- bun run typecheck:type-tests
- bun run security:pr-scan --base upstream/main --head HEAD
- bun run build
- bun run smoke

## Summary by CodeRabbit

* **New Features**
  * Added centralized message factories for constructing assistant, user, progress, interruption, tool-result stop, and model-switch breadcrumb messages.
  * Improved command-style formatting, synthetic local-command caveats, and model-switch sequences.
* **Bug Fixes**
  * Standardized handling of empty message content and improved recognition of synthetic/API-error assistant messages.
  * Refined tool-call detection limited to the last assistant turn.
* **Tests**
  * Added Bun tests validating message shapes, normalization, metadata, and breadcrumb/progress behavior for consistent outputs.

### CODE REVIEW

File: src/utils/messages/factories.ts [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+import { randomUUID, type UUID } from "crypto"
+import type { Progress } from "../../Tool.js"
+import type { SDKAssistantMessageError } from "../../entrypoints/agentSdkTypes.js"
+import type { Message, AssistantMessage, UserMessage, ProgressMessage, MessageOrigin, PartialCompactDirection } from "../../types/message.js"
+import type { PermissionMode } from "../../types/permissions.js"
+import { NO_CONTENT_MESSAGE } from "../../constants/messages.js"
+import { COMMAND_ARGS_TAG, COMMAND_MESSAGE_TAG, COMMAND_NAME_TAG, LOCAL_COMMAND_CAVEAT_TAG, LOCAL_COMMAND_STDOUT_TAG } from "../../constants/xml.js"
+import { isDangerousPermissionMode } from "../permissions/PermissionMode.js"
+
+const INTERRUPT_MESSAGE = '[Request interrupted by user]'
```
Comment (techbrewboss) [inline src/utils/messages/factories.ts]: **[Medium] Duplicated synthetic string constants create drift risk.**

These locals (`INTERRUPT_MESSAGE`, `INTERRUPT_MESSAGE_FOR_TOOL_USE`, `CANCEL_MESSAGE`, `REJECT_MESSAGE`, `NO_RESPONSE_REQUESTED`) are copied from the still-exported constants in `src/utils/messages.ts`. Factory helpers now depend on the copies, while the rest of the codebase keeps using the `messages.ts` exports.

Impact: a later edit to one copy silently desyncs synthetic detection / cancel tool-result text from the public constants.

Suggested fix: make `factories.ts` the owner and re-export from `messages.ts` (no cycle — `messages.ts` already imports this module), or land a shared constants module in this PR.

File: src/utils/messages/factories.ts [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+  messages: Message[],
+): AssistantMessage | undefined {
+  // findLast exits early from the end — much faster than filter + last for
+  // large message arrays (called on every REPL render via useFeedbackSurvey).
+  return messages.findLast(
+    (msg): msg is AssistantMessage => msg.type === 'assistant',
+  )
+}
+
+export function hasToolCallsInLastAssistantTurn(messages: Message[]): boolean {
```
Comment (techbrewboss) [inline src/utils/messages/factories.ts:81]: **[Low] Non-verbatim rewrite of `hasToolCallsInLastAssistantTurn`.**

Main still walks backward and only returns when the last assistant's `content` is an array; if that content is non-array it continues scanning earlier assistants. This version uses `getLastAssistantMessage(...)?.message.content` and returns `false` when that content is non-array.

Edge-case divergence confirmed locally (last assistant content = string, prior assistant has `tool_use`): old → `true`, new → `false`. Normal `BetaContentBlock[]` paths match.

For a pure extract PR, prefer restoring the original loop (or document the intentional behavior change).

File: src/utils/messages.ts [hunk 3]
```diff
@@ -302,349 +306,23 @@ export function buildClassifierUnavailableMessage(
...
-  toolUseID: string,
-): ToolResultBlockParam {
-  return {
-    type: 'tool_result',
-    content: CANCEL_MESSAGE,
-    is_error: true,
-    tool_use_id: toolUseID,
-  }
-}
+export {
```
Comment (techbrewboss) [inline src/utils/messages.ts:305]: Re-export surface looks correct for the moved factories. Please also fold the duplicated interrupt/cancel/reject/`NO_RESPONSE_REQUESTED` constants into this re-export block (see comment on `factories.ts`) so public `messages.ts` consumers and factory internals share one definition.

**Review by techbrewboss — CHANGES_REQUESTED** [review]
## Review summary

Part 5 of the `messages.ts` de-monolith is valuable and mostly faithful: factories move cleanly, the public barrel re-exports the right APIs, and `factories.test.ts` covers the extracted surface well. Blocking concern is duplicated synthetic string constants in `factories.ts` while `messages.ts` still exports its own copies — that is avoidable drift in a migration series. Secondary note: `hasToolCallsInLastAssistantTurn` is no longer a verbatim move (dedup rewrite) with a rare non-array content edge-case change.

## Findings

- `src/utils/messages/factories.ts:13` — **[Medium]** Local copies of `INTERRUPT_MESSAGE` / `CANCEL_MESSAGE` / `REJECT_MESSAGE` / `NO_RESPONSE_REQUESTED` (etc.) diverge from still-exported `messages.ts` constants. Own them in `factories.ts` and re-export from the barrel (or extract a shared constants module).
- `src/utils/messages/factories.ts:63` — **[Low]** `hasToolCallsInLastAssistantTurn` rewritten via `getLastAssistantMessage`; behavior matches for normal array content, but differs if the last assistant's `content` is non-array. Prefer verbatim loop for this extract, or call out the intentional change.

## Validation

- Compared extracted helpers vs `main`: factories are verbatim except `hasToolCallsInLastAssistantTurn` (rewrite) and `isSyntheticApiErrorMessage` (`export` keyword only).
- `bun test src/utils/messages/factories.test.ts` — 11 pass
- `bun run typecheck` — pass
- `bun run security:pr-scan` — clean

## Maliciousness / Risk

No malicious or supply-chain signals. Pure refactor + tests.

File: src/utils/messages/factories.ts (hunk 1 above)
Comment (jatmn) [reply src/utils/messages/factories.ts]: Addressed in `2df17ae2`: `factories.ts` now owns `INTERRUPT_MESSAGE`, `INTERRUPT_MESSAGE_FOR_TOOL_USE`, `CANCEL_MESSAGE`, `REJECT_MESSAGE`, and `NO_RESPONSE_REQUESTED`, and `SYNTHETIC_MESSAGES` is built from that single owned set. `messages.ts` imports/re-exports these values from the factories module, so there is no longer a duplicated constant source to drift.

File: src/utils/messages.ts (hunk 3 above)
Comment (jatmn) [reply src/utils/messages.ts:305]: Addressed in `2df17ae2`: the interrupt/cancel/reject/`NO_RESPONSE_REQUESTED` constants are now part of the `messages.ts` re-export block and come from `src/utils/messages/factories.ts`, so public `messages.ts` consumers and factory internals share the same definitions.

File: src/utils/messages/factories.ts (hunk 2 above)
Comment (jatmn) [reply src/utils/messages/factories.ts:81]: Addressed in `534d3227`: restored the original backwards scan behavior instead of using `getLastAssistantMessage(...)?.message.content`, so a non-array latest assistant content continues scanning earlier assistant messages as before. I also added `hasToolCallsInLastAssistantTurn preserves non-array assistant content behavior` to lock the exact edge case described here.

**Review by kevincodex1 — APPROVED** [review]
lgtm
